### PR TITLE
perf: improve KNN and ANN performance

### LINKED ISSUE #1360: perf: optimize concat_batches for vectors
As described in https://github.com/lancedb/lance/issues/1353#issuecomment-1747955084, it seems like arrow-rs doesn't pre-size buffers for the values of fixed size list. Resizing these arrays becomes a large bottleneck in KNN queries for tables with many fragments.

We should optimize this in lance, and then upstream the optimized version to arrow-rs.

### PR DESCRIPTION
**Pull request by wjones127**
* Make sure we do `take_rows` with concurrency
* Use more optimized `concat` on FSL.
* Start opening fragments ahead of reading them
* Put IO tasks in `tokio::task::spawn()` so they can make progress without the stream needing to be polled.

Closes #1360

### CODE REVIEW

**Comment by wjones127** [comment]
Results for cold start + 1,000 fragments. KNN takes 48% less time, ANN takes 80% less.
```
----------------------------------------------------------------------------------------- benchmark 'query_ann': 4 tests -----------------------------------------------------------------------------------------
Name (time in ms)                                Min                 Max                Mean             StdDev              Median                IQR            Outliers       OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_ivf_pq_index_search (NOW)                6.1702 (1.0)       13.2847 (1.0)        6.7620 (1.0)       0.9479 (1.0)        6.5009 (1.0)       0.2968 (1.0)          7;10  147.8851 (1.0)          91           1
test_ivf_pq_index_search (0001_baselin)      12.0873 (1.96)      65.2611 (4.91)      39.7912 (5.88)     11.7594 (12.41)     41.2519 (6.35)     11.5780 (39.01)        14;2   25.1312 (0.17)         46           1
test_knn_search (NOW)                       117.8859 (19.11)    126.5554 (9.53)     123.4046 (18.25)     2.6180 (2.76)     124.1701 (19.10)     2.3240 (7.83)          2;1    8.1034 (0.05)          8           1
test_knn_search (0001_baselin)              225.1686 (36.49)    266.4113 (20.05)    242.2423 (35.82)    16.3924 (17.29)    240.4475 (36.99)    24.6164 (82.93)         2;0    4.1281 (0.03)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Results for cold start + 1 fragment. KNN takes 36% less time, ANN takes 5% less time.

```
-------------------------------------------------------------------------------------- benchmark 'query_ann': 4 tests --------------------------------------------------------------------------------------
Name (time in ms)                               Min                Max               Mean            StdDev             Median               IQR            Outliers       OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_ivf_pq_index_search (0002_baselin)      2.2821 (1.0)       8.7229 (1.0)       2.7747 (1.0)      0.8354 (1.0)       2.4524 (1.0)      0.3049 (1.0)         23;32  360.4037 (1.0)         204           1
test_ivf_pq_index_search (NOW)               2.3374 (1.02)      9.5964 (1.10)      2.9449 (1.06)     0.9465 (1.13)      2.5098 (1.02)     0.7054 (2.31)        27;18  339.5739 (0.94)        220           1
test_knn_search (NOW)                       18.2445 (7.99)     33.2941 (3.82)     24.2783 (8.75)     3.6040 (4.31)     24.1712 (9.86)     3.4460 (11.30)         9;2   41.1891 (0.11)         29           1
test_knn_search (0002_baselin)              26.2865 (11.52)    51.0544 (5.85)     37.6984 (13.59)    6.0938 (7.29)     36.5303 (14.90)    6.6268 (21.74)         7;0   26.5263 (0.07)         20           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

File: rust/lance/src/dataset.rs [hunk 1]
```diff
@@ -938,13 +958,28 @@ impl Dataset {
...
-            let fragment_and_indices = fragments.iter().filter_map(|f| {
-                let local_row_ids = row_ids_per_fragment.get(&(f.id() as u64))?;
+            let fragment_and_indices = fragments.into_iter().filter_map(|f| {
+                let local_row_ids = row_ids_per_fragment.remove(&(f.id() as u64))?;
                 Some((f, local_row_ids))
             });
 
+            // manually implemented async for Send bound
+            #[allow(clippy::manual_async_fn)]
+            fn do_take(
```
Comment (eddyxu) [inline rust/lance/src/dataset.rs:968]: These two `do_take` are almost identical? Can we extract them into one?

File: rust/lance/src/index/vector/flat.rs [hunk 2]
```diff
@@ -15,15 +15,20 @@
 //! Flat Vector Index.
 //!
 
+use std::sync::Arc;
+
 use arrow::array::as_primitive_array;
 use arrow::datatypes::Float32Type;
 use arrow_array::{cast::as_struct_array, ArrayRef, RecordBatch, StructArray};
```
Comment (eddyxu) [inline rust/lance/src/index/vector/flat.rs:22]: lets' consolidate the import of same package into one line?

File: rust/lance/src/index/vector/flat.rs [hunk 3]
```diff
@@ -82,6 +57,8 @@ pub async fn flat_search(
         return Ok(RecordBatch::new_empty(schema_with_distance.into()));
     }
 
+    // TODO: waiting to do this until the end adds quite a bit of latency. We should
```
Comment (eddyxu) [inline rust/lance/src/index/vector/flat.rs:60]: IIUC, waiting here is the existing behavior, there is no regression, right?

File: rust/lance/src/io/exec/take.rs [hunk 4]
```diff
@@ -108,6 +98,27 @@ impl Take {
             output_schema,
         }
     }
+
+    // manually implemented async for Send bound
+    #[allow(clippy::manual_async_fn)]
+    fn take_batch(
+        batch: RecordBatch,
```
Comment (eddyxu) [inline rust/lance/src/io/exec/take.rs:105]: Should we just pass row_id array, which makes this function more self-contained.

File: rust/lance/src/index/vector/flat.rs (hunk 3 above)
Comment (wjones127) [reply rust/lance/src/index/vector/flat.rs:60]: Yeah, this is existing behavior. Just adding in more details from observations in tracing.

File: rust/lance/src/io/exec/take.rs (hunk 4 above)
Comment (wjones127) [reply rust/lance/src/io/exec/take.rs:105]: `row_ids` would have to be a reference to a column in `batch`, and it's much easier if all the parameters have `'static` lifetime.

I added a docstring which I think makes sense as a unit of work.

**Comment by wjones127** [comment]
@eddyxu Rereview?

**Review by eddyxu — APPROVED** [review]

**Pull request merged**